test/instruction_graph_reduction_tests.cc

```diff
@@ -0,0 +1,485 @@
+	ictx.finish();
+
+	const auto all_instrs = ictx.query_instructions();
+	const auto all_pilots = ictx.query_outbound_pilots();
+
+	// there is exactly one (global) reduce-instruction per node.
+	const auto reduce = all_instrs.select_unique<reduce_instruction_record>();
+	CHECK(reduce->scope == reduce_instruction_record::reduction_scope::global);
+	CHECK(reduce->num_source_values == num_nodes);
+	CHECK(reduce->buffer_id == buf.get_id());
+	CHECK(reduce->box == box<3>(zeros, ones));
+
+	// we send partial results to the peer - this operation anti-depends on the reduce operation, which will overwrite its buffer
+	const auto send_to_peer = all_instrs.select_unique<send_instruction_record>();
+	CHECK(reduce.predecessors().contains(send_to_peer));
+	CHECK(send_to_peer->offset_in_buffer == zeros);
+	CHECK(send_to_peer->send_range == ones);
+	CHECK(send_to_peer->dest_node_id == peer_nid);
+	CHECK(send_to_peer->transfer_id.rid == reduce->reduction_id);
+	CHECK(send_to_peer->transfer_id.bid == buf.get_id());
+
+	// fill the gather-buffer before initiating the gather-receive because if the peer decides to not send a payload (but an empty pilot), the gather-recv can
+	// simply skip writing to the appropriate position in the gather buffer.
+	const auto fill_identity = all_instrs.select_unique<fill_identity_instruction_record>();
+	CHECK(fill_identity->reduction_id == reduce->reduction_id);
+	CHECK(fill_identity->num_values == num_nodes);
+
+	// the global gather buffer is a single allocation
+	const auto gather_alloc = fill_identity.predecessors().select_unique<alloc_instruction_record>();
+	CHECK(gather_alloc->origin == alloc_instruction_record::alloc_origin::gather);
+	CHECK(gather_alloc->size_bytes == num_nodes * sizeof(int));
+	CHECK(gather_alloc->allocation_id == fill_identity->allocation_id);
+
+	// we (gather-) copy the local partial result to the appropriate position in the gather buffer
+	const auto gather_copy = reduce.predecessors().select_unique<copy_instruction_record>();
+	CHECK(gather_copy->origin == copy_instruction_record::copy_origin::gather);
+	CHECK(gather_copy->source_allocation.offset_bytes == 0);
+	CHECK(gather_copy->dest_allocation.offset_bytes == local_nid * sizeof(int));
+	CHECK(gather_copy->copy_region == region(box<3>(zeros, ones)));
+
+	// we gather-receive from all peers - this will _not_ write to the position `local_nid`
+	const auto gather_recv = all_instrs.select_unique<gather_receive_instruction_record>();
+	CHECK(reduce.predecessors().contains(gather_recv));
+	CHECK(gather_recv->gather_box == box<3>(zeros, ones));
+	CHECK(gather_recv->num_nodes == num_nodes);
+	CHECK(gather_recv->allocation_id == gather_copy->dest_allocation.id);
+	CHECK(gather_recv->transfer_id == send_to_peer->transfer_id);
+}
+
+TEST_CASE("reduction accesses on a multi-node multi-device setup generate global and local reduce-instructions",
+    "[instruction_graph_generator][instruction-graph][reduction]") //
+{
+	const size_t num_nodes = 2;
```

Comment:
Should this be tested on more than 2 nodes? (Since that's arguably still a special case)

This might get really unwieldy on the graph though, I guess a better option is to functionally test more complex reduction scenarios as integration tests (when we have execution).

Reply:
I have extended these test cases to run both with 2 and 4 nodes.